src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsEntityTypeDefinition.cs

```diff
@@ -112,7 +112,36 @@ private IEnumerable<IEdmStructuralProperty> ComputeDeclaredKey()
                 List<IEdmStructuralProperty> key = new List<IEdmStructuralProperty>();
                 foreach (CsdlPropertyReference keyProperty in this.entity.Key.Properties)
                 {
-                    IEdmStructuralProperty structuralProperty = this.FindProperty(keyProperty.PropertyName) as IEdmStructuralProperty;
+                    IEdmStructuralProperty structuralProperty = null;
+                    if (!string.IsNullOrEmpty(keyProperty?.Alias))
+                    {
+                        string[] keyPropertySegments = keyProperty.PropertyName.Split('/');
+                        if (keyPropertySegments.Length > 0)
+                        {
+                            structuralProperty = this.FindProperty(keyPropertySegments[0]) as IEdmStructuralProperty;
+                            for (int i = 1; i < keyPropertySegments.Length; i++)
+                            {
+                                if (structuralProperty.Type.Definition.TypeKind == EdmTypeKind.Complex)
+                                {
+                                    structuralProperty = structuralProperty.Type.AsComplex().FindProperty(keyPropertySegments[i]) as IEdmStructuralProperty;
+                                }
+                                else if (structuralProperty.Type.Definition.TypeKind == EdmTypeKind.Entity)
+                                {
+                                    structuralProperty = structuralProperty.Type.AsEntity().FindProperty(keyPropertySegments[i]) as IEdmStructuralProperty;
+                                }
+                                else
+                                {
+                                    key.Add(new UnresolvedProperty(this, keyProperty.PropertyName, this.Location));
+                                }
+                            }
+                            structuralProperty = new EdmStructuralPropertyAlias(structuralProperty.DeclaringType, structuralProperty.Name, structuralProperty.Type, keyProperty.Alias, keyPropertySegments);
+                        }
+                    }
+                    else
+                    {
+                        structuralProperty = this.FindProperty(keyProperty.PropertyName) as IEdmStructuralProperty;
+                    }

```

Comment:
The nested `if` above means that we won't always reach this branch even if we should fallback to the existing behavior. 

```suggestion
                    structuralProperty = structuralProperty ?? this.FindProperty(keyProperty.PropertyName) as IEdmStructuralProperty;
```